### PR TITLE
Introduced complementary palette

### PR DESCRIPTION
**Pull request by strouhaljan**
- with auto-generated missing shades
- rendered as css variables: --gd-palette-complementary-<0-9>

JIRA: BB-2813

---

Supported PR commands:

| Command                | Description            |
| ---------------------- | ---------------------- |
| `ok to test`           | Re-run standard checks |
| `extended test`        | BackstopJS tests       |
| `extended check sonar` | SonarQube tests        |

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `check-extended` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2536/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/4166/ : SUCCESS in 14m 56s

File: libs/sdk-ui-theme-provider/src/complementaryPalette.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    | "shade0"
+    | "shade1"
+    | "shade2"
+    | "shade3"
+    | "shade4"
+    | "shade5"
+    | "shade6"
+    | "shade7"
+    | "shade8"
+    | "shade9";
```
Comment (lupko) [inline libs/sdk-ui-theme-provider/src/complementaryPalette.ts:17]: i wonder can you use keyof here and have the keys derived from the type?

File: libs/sdk-ui-theme-provider/src/tests/complementaryPalette.test.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+                    shade1: "#32321b",
+                    shade2: "#645415",
+                    shade3: "#97760f",
+                    shade4: "#c99809",
+                    shade5: "#fcba03",
+                    shade6: "#bfa306",
+                    shade7: "#828c09",
+                    shade8: "#45750c",
+                    shade9: "#095f0f",
+                },
```
Comment (lupko) [inline libs/sdk-ui-theme-provider/src/tests/complementaryPalette.test.ts:37]: i suggest to switch this to snapshot test. let the computer do the 'typing' of the expected value for you.

File: libs/sdk-backend-spi/src/workspace/styling/theme.ts [hunk 3]
```diff
@@ -92,6 +110,11 @@ export interface IThemePalette {
      * Used to express info or progress
      */
     info?: IThemeColorFamily;
+
+    /**
+     * Complementary palette
+     */
```
Comment (lupko) [inline libs/sdk-backend-spi/src/workspace/styling/theme.ts:116]: this doc should be expanded to describe what is the complementary palette used for. what can be achieved using it.

File: libs/sdk-ui-theme-provider/src/tests/complementaryPalette.test.ts (hunk 2 above)
Comment (strouhaljan) [reply libs/sdk-ui-theme-provider/src/tests/complementaryPalette.test.ts:37]: Good point, thanks.

File: libs/sdk-ui-theme-provider/src/complementaryPalette.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        shade2: palette.shade2 || getShade(palette, 2),
+        shade3: palette.shade3 || getShade(palette, 3),
+        shade4: palette.shade4 || getShade(palette, 4),
+        shade5: palette.shade5 || getShade(palette, 5),
+        shade6: palette.shade6 || getShade(palette, 6),
+        shade7: palette.shade7 || getShade(palette, 7),
+        shade8: palette.shade8 || getShade(palette, 8),
+        shade9: palette.shade9,
+    };
+};
```
[the remark this answers is not in the document]
Comment (strouhaljan) [reply libs/sdk-ui-theme-provider/src/complementaryPalette.ts:54]: Yeah, it is :D I should mention this is the second version of the algorithm. The first one had everything in two nested reducers, but it looked waaay too complicated. Than I came up with a bit easier-to-understand version, where the shades are listed in return of `getComplementaryPalette()` (it's only 10 of them and we do not plan to extend it so far).

File: libs/sdk-ui-theme-provider/src/complementaryPalette.ts (hunk 1 above)
Comment (strouhaljan) [reply libs/sdk-ui-theme-provider/src/complementaryPalette.ts:17]: Absolutely.

File: libs/sdk-backend-spi/src/workspace/styling/theme.ts (hunk 3 above)
Comment (strouhaljan) [reply libs/sdk-backend-spi/src/workspace/styling/theme.ts]: Makes sense. Take a look at force-pushed changes now.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2537/ : SUCCESS in 11s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/4168/ : SUCCESS in 15m 36s

**Review by no23reason — APPROVED** [review]
LGTM

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by yenkins** [comment]
### Sonar scan result
 - No new bug found. Congratulation!

More detail, see in https://sonarqube-dev.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR1054

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/764/ : SUCCESS in 25s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/1027/ : SUCCESS in 17s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/4171/ : SUCCESS in 22m 59s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-el7/126/ : SUCCESS in 10m 40s (non-voting)

**Pull request merged**